### ISSUE TITLE
Some short lines added with `plot()` are missing in the PNG output depending on DPI

### ISSUE THREAD

**Issue by tskir** [matplotlib/matplotlib#20192]
### Bug report

Depending on DPI, a figure saved as PNG is missing some short lines.

**Code for reproduction**

```python
import matplotlib
from matplotlib import pyplot as plt
import numpy as np

matplotlib.use('agg')

# Generate a set of 50 random points.
np.random.seed(0)
all_x, all_y = (np.random.rand(50) for _ in range(2))

# Scatter dots in blue.
plt.scatter(all_x, all_y, c='blue', s=2)

# Plot very short lines (length 0.001) in red with a small offset on Y.
for x, y in zip(all_x, all_y):
    plt.plot((x, x + 0.001), (y + 0.01, y + 0.01), c='red', linewidth=2)

# Save with different resolutions
plt.savefig('bug_96.png', dpi=96)
plt.savefig('bug_300.png', dpi=300)
```

**Actual outcome**

Depending on export DPI, some short segments (in red, added by `plot()`) are missing, while all dots (in blue, added by `scatter()` are always present).

At 96 DPI:
![image](https://user-images.githubusercontent.com/10669118/117580746-64d1a580-b0f1-11eb-9c75-d7ca028d230b.png)

At 300 DPI:
![image](https://user-images.githubusercontent.com/10669118/117580753-6ac78680-b0f1-11eb-83df-d0701b0dba1d.png)

**Expected outcome**

The segments should be always present, regardless of the resolution.

**Matplotlib version**

  * Operating system: Ubuntu 20.04.2 LTS
  * Matplotlib version (`import matplotlib; print(matplotlib.__version__)`): 3.4.2
  * Matplotlib backend (`print(matplotlib.get_backend())`): `agg`
  * Python version: 3.8.5

Python is installed via apt-get, Matplotlib installed system-wide using PIP.

**Comment by tskir** [comment]
Interestingly enough, the bug disappears with `matplotlib.use('cairo')`.

At 96 DPI with cairo:
![image](https://user-images.githubusercontent.com/10669118/117581039-a9117580-b0f2-11eb-9700-bc833ead3ba0.png)

At 300 DPI with cairo:
![image](https://user-images.githubusercontent.com/10669118/117581042-aca4fc80-b0f2-11eb-9623-926c56349b07.png)

**Comment by jklymak** [comment]
If you turn snapping off this is fixed.  `plot(....snap=False)`

**Comment by jklymak** [comment]
FWIW, I think snapping for ticks makes some sense.  I don't think there are many situations where having snapping on for other artists is a good idea at all.  I'm not familiar enough with matplotlib's history wrt snapping, but it seems a bad idea in general.  

I also recommend strongly getting around any bad ideas matplotlib may have had by saving at relatively high dpi and then using a graphics package to reduce the size optically.  In my opinion we try to do too many "smart" things with low resolution images that really should be done in post-processing, not while we compose the images.

**Comment by tskir** [comment]
@jklymak Indeed, thank you very much for the pointer and for such a quick reply!

I'll apply this fix, and will leave this issue up for any related discussions.

**Comment by tacaswell** [comment]
I have somewhat mixed thoughts on this.  On one hand, snapping things out of existence is bad, but on the other what is getting rendered in this case is relatively miss-leading 

The snapping code only kicks in on paths if the lines are only horizontal or vertical which this case does hit.  This is to make sure a bunch of long horizontal lines all look the same weight.  Think of a 2 pixel wide line, if the path goes down the boundary between two pixels you get a 2 pixel wide 100% color rendering.  If the path goes down the center of a pixel, then you get a 3 pixel wide line with the center row being 100% and the outer pixels being 50% (due to anti-aliasing) or a 3 pixel wide line (if you turn off AA). These defaults were decided back before high-dpi screens existed so (what we now call) low-dpi was pretty important.  

On the other hand, what is being rendered for the lines is mostly the caps not the actual line (setting `solid_capstyle='butt'` shows this), however in that case we are still seeing the same issue.

I think the right fix is to add logic in https://github.com/matplotlib/matplotlib/blob/9b61b471d0a5bcadbb68a35515e7598b9640c9bf/src/path_converters.h#L559-L599 that turns off snapping for path segments that are 2 points long and are less than 2 pixel apart.

This patch seems to fix the problem, but have not tested what other consequences this has.

```diff
diff --git a/src/path_converters.h b/src/path_converters.h
index 1482aeed95..e8275974fa 100644
--- a/src/path_converters.h
+++ b/src/path_converters.h
@@ -561,8 +561,9 @@ class PathSnapper
         // If this contains only straight horizontal or vertical lines, it should be
         // snapped to the nearest pixels
         double x0 = 0, y0 = 0, x1 = 0, y1 = 0;
+        double xstart, ystart;
         unsigned code;
-
+        int lineto_count = 0;
         switch (snap_mode) {
         case SNAP_AUTO:
             if (total_vertices > 1024) {
@@ -570,6 +571,8 @@ class PathSnapper
             }
 
             code = path.vertex(&x0, &y0);
+            xstart = x0;
+            ystart = y0;
             if (code == agg::path_cmd_stop) {
                 return false;
             }
@@ -580,14 +583,20 @@ class PathSnapper
                 case agg::path_cmd_curve4:
                     return false;
                 case agg::path_cmd_line_to:
+                    lineto_count++;
                     if (fabs(x0 - x1) >= 1e-4 && fabs(y0 - y1) >= 1e-4) {
                         return false;
                     }
                 }
                 x0 = x1;
                 y0 = y1;
-            }
 
+            }
+            // we only have two verticies and the displacement in either direction
+            // is less than two pixels
+            if (lineto_count == 1 && (fabs(xstart - x0) < 2 && fabs(ystart - y0) < 2)) {
+                return false;
+            }
             return true;
         case SNAP_FALSE:
             return false;

```
